### PR TITLE
feat: package up dnsmasq as a DNS and DHCP server

### PR DESCRIPTION
**Pull request by skrobul**
This PR adds dnsmasq installation to the `openstack` namespace. It will:
- hand out IP addresses to hosts that are provisioned with Ironic
- cooperate with Ironic for setting per-host options
- enable a DNS server that can resolve records for endpoint URLs returned by the OpenStack Identity service catalogue

In its current form, it supports only a single DHCP range/pool. Support for multiple pools will be added as a separate story as it requires more input parameters (MACs of the relays, range names, and potentially different DNS server IPs).

Closes PUC-225

### CODE REVIEW

**Comment by cardoe** [comment]
Probably some commits can get squashed into others but overall I think it's good. Need to stop systemd-resolvd or whatever it's called if you use container hostPort 53 however. We can have it listen on the ingress if you configure the ingress to listen on UDP 53. We can also have it listen on a service IP.

**Review by nicholaskuechler — APPROVED** [review]

File: containers/Dockerfile.dnsmasq [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+COPY common/helpers.sh /helpers.sh
+COPY dnsmasq/entry-point.sh /entry-point.sh
+RUN chmod +x /entry-point.sh
+COPY dnsmasq/dnsmasq.conf.j2 /etc/dnsmasq.conf.j2
+# let our entry point write out the script
+RUN ln -sf /etc/dnsmasq.d/dnsmasq.conf /etc/dnsmasq.conf
+
+EXPOSE 53/udp
+EXPOSE 67/udp
```
Comment (cardoe) [inline containers/Dockerfile.dnsmasq:14]: Probably should drop the port 53 since this PR makes it listen on 1053.

Comment (cardoe) [reply containers/Dockerfile.dnsmasq:14]: And we listen on port 1067 for DHCP

File: .github/workflows/containers.yaml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        uses: docker/metadata-action@v5
+        with:
+          images: ghcr.io/rackerlabs/openstackhelm/ironic
+          tags: |
+            type=raw,value=latest,enable={{is_default_branch}}
+            type=sha,enable={{is_default_branch}}
+            type=raw,value=2023.1-ubuntu_jammy
+          labels: |
+            org.opencontainers.image.title=ironic with ISO utils
+            org.opencontainers.image.base.name=ghcr.io/rackerlabs/openstackhelm/ironic
```
Comment (cardoe) [inline .github/workflows/containers.yaml:42]: So that's not really correct. This should be the container we're calling `FROM` in the `Dockerfile`, no?

Comment (skrobul) [reply .github/workflows/containers.yaml]: Good catch, thanks ! fixed in the rebase.

File: containers/Dockerfile.dnsmasq (hunk 1 above)
Comment (skrobul) [reply containers/Dockerfile.dnsmasq]: fixed in the rebase

**Pull request merged**